src/uct/ib/mlx5/ib_mlx5_md.c

```diff
@@ -0,0 +1,889 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2001-2018.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#include "ib_mlx5.h"
+#include "ib_mlx5_log.h"
+#include "ib_mlx5_ifc.h"
+
+#include <ucs/arch/bitops.h>
+#include <ucs/profile/profile.h>
+
+typedef struct uct_ib_mlx5_mem {
+    uct_ib_mem_t             super;
+#if HAVE_DEVX
+    struct mlx5dv_devx_obj   *atomic_dvmr;
+#elif HAVE_EXP_UMR
+    struct ibv_mr            *atomic_mr;
+#endif
+} uct_ib_mlx5_mem_t;
+
+typedef struct uct_ib_mlx5_dbrec_page {
+    struct mlx5dv_devx_umem *mem;
+} uct_ib_mlx5_dbrec_page_t;
+
+
+uint8_t uct_ib_mlx5_md_get_atomic_mr_id(uct_ib_mlx5_md_t *md)
+{
+#if HAVE_EXP_UMR
+    if ((md->umr_qp == NULL) || (md->umr_cq == NULL)) {
+        return 0;
+    }
+    /* Generate atomic UMR id. We want umrs for same virtual addresses to have
+     * different ids across processes.
+     *
+     * Usually parallel processes running on the same node as part of a single
+     * job will have consecutive PIDs. For example MPI ranks, slurm spawned tasks...
+     */
+    return getpid() % 256;
+#else
+    return 0;
+#endif
+}
+
+static uct_ib_md_ops_t uct_ib_mlx5dv_md_ops;
+
+#if HAVE_DEVX
+
+static ucs_status_t uct_ib_mlx5dv_memh_reg(uct_ib_md_t *ibmd,
```

Comment:
uct_ib_mlx5_devx_reg_atomic_key
and have different ops structure for devx / exp